kernel/eventfddev.c

```diff
@@ -0,0 +1,537 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/eventfd.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <myst/cond.h>
+#include <myst/defs.h>
+#include <myst/eraise.h>
+#include <myst/eventfddev.h>
+#include <myst/id.h>
+#include <myst/panic.h>
+#include <myst/process.h>
+#include <myst/round.h>
+#include <myst/syscall.h>
+
+#define MAGIC 0x9906acdc
+
+#define MAX_COUNTER_VALUE ((uint64_t)0xfffffffffffffffe)
+
+struct myst_eventfd
+{
+    uint32_t magic;
+    int flags;
+    int fdflags;
+    uint64_t counter;
+    myst_mutex_t mutex;
+    myst_cond_t cond;
+};
+
+MYST_INLINE bool _valid_eventfd(const myst_eventfd_t* eventfd)
+{
+    return eventfd && eventfd->magic == MAGIC;
+}
+
+static void _lock(myst_eventfd_t* eventfd)
+{
+    myst_assume(_valid_eventfd(eventfd));
+    myst_mutex_lock(&eventfd->mutex);
+}
+
+static void _unlock(myst_eventfd_t* eventfd)
+{
+    myst_assume(_valid_eventfd(eventfd));
+    myst_mutex_unlock(&eventfd->mutex);
+}
+
+static int _eventfd(
+    myst_eventfddev_t* eventfddev,
+    unsigned int initval,
+    int flags,
+    myst_eventfd_t** eventfd_out)
+{
+    int ret = 0;
+    myst_eventfd_t* eventfd = NULL;
+
+    if (!eventfddev || !eventfd_out)
+        ERAISE(-EINVAL);
+
+    if ((flags & ~(EFD_CLOEXEC | EFD_NONBLOCK | EFD_SEMAPHORE)))
+        ERAISE(-EINVAL);
+
+    /* Create the read eventfd */
+    {
+        if (!(eventfd = calloc(1, sizeof(myst_eventfd_t))))
+            ERAISE(-ENOMEM);
+
+        eventfd->magic = MAGIC;
+        eventfd->counter = initval;
+
+        if (flags & EFD_CLOEXEC)
+            eventfd->fdflags = FD_CLOEXEC;
+
+        eventfd->flags = flags;
+    }
+
+    *eventfd_out = eventfd;
+    eventfd = NULL;
+
+done:
+
+    if (eventfd)
+        free(eventfd);
+
+    return ret;
+}
+
+static ssize_t _read(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    void* buf,
+    size_t count)
+{
+    ssize_t ret = 0;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EBADF);
+
+    if (!buf)
+        ERAISE(-EINVAL);
+
+    if (count < sizeof(uint64_t))
+        ERAISE(-EINVAL);
+
+    _lock(eventfd);
+
+    if (!_valid_eventfd(eventfd))
+    {
+        /* cannot unlock since mutex is no longer valid */
+        ERAISE(-EBADF);
+    }
+
+    /* if the counter is readable (non-zero) */
+    if (eventfd->counter != 0)
+    {
+        memcpy(buf, &eventfd->counter, sizeof(uint64_t));
+
+        if ((eventfd->flags & EFD_SEMAPHORE))
+            eventfd->counter--;
+        else
+            eventfd->counter = 0;
+
+        myst_cond_signal(&eventfd->cond);
+        _unlock(eventfd);
```

Comment:
Why don't we move `_unlock` to "done" section?

Reply:
Good idea. Fixed.